### PR TITLE
Fix infantry staying out of range

### LINKED ISSUE #5841: Rocket infantry stays out of firing range
I noticed that when you order a TD rocket trooper to fire on te ground outside of its weapon range, it will only move closer but not start firing. Looks like it stays about 1 cell out of weapon range.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/2985845-rocket-infantry-stays-out-of-firing-range?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by reaperrr**
Closes #5841.

### CODE REVIEW

**Comment by reaperrr** [comment]
Should go into playtest, in my opinion.
It's an annoying, long-standing bug and a relatively simple fix.

**Review by abcdefg30 — APPROVED** [review]

**Comment by forcecore** [comment]
Now I see why infantries are staring at each other in AI vs AI games

File: OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs [hunk 1]
```diff
@@ -47,8 +47,13 @@ protected override IEnumerable<CPos> CandidateMovementCells(Actor self)
...
 			var minCells = minRange.Length / 1024;
 
+			// AtCorrectRange(map.CenterOfCell(c)) will return the current cell if the cell center is in range,
+			// even if the actor is actually out of range due to its current subcell not being the center.
+			// Avoid that by not including the current Location if the current actor.CenterPosition is out of range.
+			var ignoreCurrentCell = !AtCorrectRange(self.CenterPosition);
+
 			return map.FindTilesInAnnulus(targetPosition, minCells, maxCells)
-				.Where(c => AtCorrectRange(map.CenterOfCell(c)));
+				.Where(c => !(ignoreCurrentCell && c == self.Location) && AtCorrectRange(map.CenterOfCell(c)));
```
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs]: A more correct solution here is to use `map.CenterOfSubCell(c, subCell)`, where you pass `subCell` in via `CandidateMovementCells(self, Mobile.FromSubCell);`

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs]: > where you pass subCell in via CandidateMovementCells(self, Mobile.FromSubCell);

I wasn't sure how you meant that, so for I've simply gone for
`c => AtCorrectRange(map.CenterOfSubCell(c, Mobile.FromSubCell))`.

**Comment by reaperrr** [comment]
Updated, but left the fixup as separate commit for now in case it needs more changes.

File: OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs (hunk 1 above)
Comment (pchote) [inline OpenRA.Mods.Common/Activities/Move/MoveWithinRange.cs]: Ah right, I missed that `Mobile` was already available from that context.

**Comment by reaperrr** [comment]
Squashed commits.

**Comment by abcdefg30** [comment]
Still works, and my 👍 still stands.

**Review by GraionDilach — APPROVED** [review]
LGTM. I consider this a must for the playtest tbh.

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ba6243e86eea2bd06ce4b4625d460fc825268190...167ee379a3802e8e00b4e6b079d647b65cd578ff

**Pull request merged**